pkg/spdk/replica.go

```diff
@@ -1543,6 +1609,558 @@ func (r *Replica) SnapshotRangeHashGet(spdkClient *spdkclient.Client, snapshotNa
+			r.log.WithError(err).Errorf("Failed to release the cloning port %d for cloning dst cleanup", r.snapshotCloningDstCache.cloningPort)
+			aggregatedErrors = append(aggregatedErrors, err)
+		} else {
+			r.snapshotCloningDstCache.cloningPort = 0
+			r.snapshotCloningDstCache.cloningLvolAddress = ""
+		}
+	}
+	if _, err := spdkClient.BdevLvolDelete(spdktypes.GetLvolAlias(r.LvsName, cloningLvolName)); err != nil && !jsonrpc.IsJSONRPCRespErrorNoSuchDevice(err) {
+		r.log.WithError(err).Errorf("Failed to delete the cloning lvol %s for cloning dst cleanup", cloningLvolName)
+		aggregatedErrors = append(aggregatedErrors, err)
+	} else {
+		r.snapshotCloningDstCache.cloningLvol = nil
+	}
+
+	if _, err := spdkClient.BdevLvolDelete(spdktypes.GetLvolAlias(r.LvsName, cloningLvolName)); err != nil && !jsonrpc.IsJSONRPCRespErrorNoSuchDevice(err) {
+		r.log.WithError(err).Errorf("Failed to delete the cloning lvol %s for cloning dst cleanup", cloningLvolName)
+		aggregatedErrors = append(aggregatedErrors, err)
+	} else {
+		r.snapshotCloningDstCache.cloningLvol = nil
+	}
+
+	tmpSnapName := GetTmpSnapNameForCloningLvol(r.Name)
+	if _, err := spdkClient.BdevLvolDelete(spdktypes.GetLvolAlias(r.LvsName, tmpSnapName)); err != nil && !jsonrpc.IsJSONRPCRespErrorNoSuchDevice(err) {
+		r.log.WithError(err).Errorf("Failed to delete the tmp snapshot %s for cloning dst cleanup", cloningLvolName)
+		aggregatedErrors = append(aggregatedErrors, err)
+	}
+
+	r.snapshotCloningDstCache.srcReplicaName = ""
+	r.snapshotCloningDstCache.srcReplicaAddress = ""
+	if r.snapshotCloningDstCache.monitorCancelFunc != nil {
+		r.snapshotCloningDstCache.monitorCancelFunc()
+		r.snapshotCloningDstCache.monitorCancelFunc = nil
+	}
+
+	if clearStatus {
+		r.snapshotCloningDstCache.processedClusters = 0
+		r.snapshotCloningDstCache.totalClusters = 0
+		r.snapshotCloningDstCache.cloningError = ""
+		r.snapshotCloningDstCache.cloningState = ""
+	}
+
+	return util.CombineErrors(aggregatedErrors...)
+}
+
+func (r *Replica) snapshotLinkedCloneSrcStart(spdkClient *spdkclient.Client, snapshotName, dstReplicaName string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "failed to do snapshotLinkedCloneSrcStart")
+	}()
+
+	bdevLvolMap, err := GetBdevLvolMapWithFilter(spdkClient, r.replicaLvolFilter)
+	if err != nil {
+		return err
+	}
+
+	existingParentOfDstReplica := ""
+	for lvolName, lvol := range bdevLvolMap {
+		if types.IsBackingImageSnapLvolName(lvolName) {
+			continue
+		}
+		for _, childLvolName := range lvol.DriverSpecific.Lvol.Clones {
+			if childLvolName == dstReplicaName {
+				existingParentOfDstReplica = lvolName
+				continue
+			}
+			if !IsReplicaLvol(r.Name, childLvolName) {
+				return fmt.Errorf("there are already another linked-clone lvol %v in src replica %v. "+
+					"Each src replica can only has 1 linked-clone lvol at a time", childLvolName, r.Name)
+			}
```

Comment:
We will rely on the upper layer (longhorn-manager) checking if other replicas of the src volume have a linked-clone, right?